src/float.jl

```diff
@@ -0,0 +1,352 @@
+
+The IEEE 754 bias of the exponent bits of a floating point number of type `F`.
+"""
+bias(::Type{T}) where {T<:Base.IEEEFloat} = uint(T)(1 << (exposize(T) - 1) - 1)
+
+function masks(::Type{T}) where {T<:Base.IEEEFloat}
+    ui = uint(T)
+    signbitmask = one(ui) << (8 * sizeof(ui) - 1)
+    fracbitmask = (-1 % ui) >> (8 * sizeof(ui) - fracsize(T))
+    expobitmask = ((-1 % ui) >> (8 * sizeof(ui) - exposize(T))) << fracsize(T)
+    signbitmask, expobitmask, fracbitmask
+end
+
+"""
+    assemble(::T, sign::I, expo::I, frac::I) where {I, T <: Union{Float16, Float32, Float64}} -> T
+
+Assembles `sign`, `expo` and `frac` arguments into the floating point number of type `T` it represents.
+`sizeof(T)` must match `sizeof(I)`.
+"""
+function assemble(::Type{T}, sign::I, expo::I, frac::I) where {I,T<:Base.IEEEFloat}
+    sizeof(T) == sizeof(I) || throw(ArgumentError("The bitwidth of  `$T` needs to match the other arguments of type `I`!"))
+    signmask, expomask, fracmask = masks(T)
+    sign = (sign << (exposize(T) + fracsize(T))) & signmask
+    expo = (expo << fracsize(T)) & expomask
+    frac = frac & fracmask
+    ret = sign | expo | frac
+    reinterpret(T, ret)
+end
+
+"""
+    tear(x::T) where T <: Union{Float16, Float32, Float64} -> Tuple{I, I, I}
+
+Returns the sign, exponent and fractional parts of a floating point number.
+The returned tuple consists of three unsigned integer types `I` of the same bitwidth as `T`.
+"""
+function tear(x::T) where {T<:Base.IEEEFloat}
+    signbitmask, expobitmask, fracbitmask = masks(T)
+    ur = reinterpret(uint(T), x)
+    s = (ur & signbitmask) >> (exposize(T) + fracsize(T))
+    e = (ur & expobitmask) >> fracsize(T)
+    f = (ur & fracbitmask) >> 0x0
+    s, e, f
+end
+
+"""
+    exponent_key(T, e)
+
+A lexicographical ordering for floating point exponents. The encoding is ported
+from Hypothesis.
+
+The ordering is
+
+    * non-negative exponents in increasing order
+    * negative exponents in decreasing order
+    * the maximum exponent
+
+# Extended Help
+
+The [reference implementation](https://github.com/HypothesisWorks/hypothesis/blob/aad70fb2d9dec2cef9719cdf5369eec9fae0d2a4/hypothesis-python/src/hypothesis/internal/conjecture/floats.py#L82) in Hypothesis.
+"""
+function exponent_key(::Type{T}, e::iT) where {T<:Base.IEEEFloat,iT<:Unsigned}
+    if e == max_exponent(T)
+        return Inf
+    end
+    unbiased = float(e) - bias(T)
+    if unbiased < 0
+        # order all negative exponents after the positive ones
+        # in reverse order
+        # max_exponent(T) - 1 maps to the key bias(T)
+        # so the first negative exponent maps to bias(T) + 1
+        bias(T) - unbiased
+    else
+        unbiased
+    end
+end
+
+""" 
+    _make_encoding_table(T)
+
+Build a look up table for encoding exponents of floating point numbers of type `T`.
+For a floating point type `T`, the lookup table is a permutation of the unsigned integers of type [`uint`](@ref) from `0` to `max_exponent(T)`.
```

Comment:
```suggestion
For a floating point type `T`, the lookup table is a permutation of the unsigned integers of type [`uint(T)`](@ref) from `0` to `max_exponent(T)`.
```

We'll have to see whether Documenter likes the added parameter. If it doesn't, we can drop the `@ref`.